File: doc/src/codes/Users_Manual/advanced_examples/custom_gain_0.py
import numpy as np
from pyautd3 import Drive, EmitIntensity, Geometry, Phase
from pyautd3.gain import Gain


class Focus(Gain):
    def __init__(self, point):
        self.point = np.array(point)

    def calc(self, geometry: Geometry) -> dict[int, np.ndarray]:
        return Gain._transform(
            geometry,
            lambda dev, tr: Drive(
                Phase.from_rad(
                    np.linalg.norm(tr.position - self.point)
                    * tr.wavenumber(dev.sound_speed)
                ),
                EmitIntensity.maximum(),
            ),
        )

File: doc/src/codes/Users_Manual/advanced_examples/custom_gain_0.cs
public class FocalPoint : Gain
{
    private readonly Vector3d _point;

    public FocalPoint(Vector3d point)
    {
        _point = point;
    }

    public override Dictionary<int, Drive[]> Calc(Geometry geometry)
    {
        return Transform(geometry, (dev, tr) =>
        {
            var tp = tr.Position;
            var dist = (tp - _point).L2Norm;
            var phase = dist * tr.Wavenumber(dev.SoundSpeed);
            return new Drive { Phase = Phase.FromRad(phase), Intensity = EmitIntensity.Max };
        });
    }
}